### PR TITLE
linker: sort app shared mem partition by alignment

### LINKED ISSUE #14121: gaps between app shared memory partitions can waste a lot of space
**Describe the bug**

On systems which require power-of-two size/alignment, we are not ordering the app shared memory partitions in an optimal way, which results in a lot of memory being wasted.

A good example is running the app memory size script from #14120 on the kernel produced by tests/subsys/jwt on mps2_an385 after applying #14115. Each partition is either very small or has good efficiency, but notice the huge gap in between z_libc_partition and k_mbedtls_partition:

```
$ python3 ~/projects/zephyr4/scripts/appmem_report.py zephyr/zephyr.elf 
Userspace memory partition report:
Partition                     Total       Used        Wasted      Efficiency
----------------------------------------------------------------------------
ztest_mem_partition           0x40        0x34        0xc         81.25%
  gap 0x0         
z_libc_partition              0x20        0x4         0x1c        12.50%
  gap 0x7fa0      
k_mbedtls_partition           0x8000      0x7fe0      0x20        99.90%

total                         0x10000     0x8018      0x7fe8      50.04%
```

Ordering the app shared memory partitions in reverse order of total size should greatly reduce these gaps.

We don't know the full sizes of these partitions when gen_app_partitions.py runs since the zephyr_prebuilt.elf has not been linked yet, but since we are examining the individual object files we might be able to assemble a total anyway. Or we could somehow figure out a way to get the linker to reorganize these partitions for us, but that might require them being placed in individual sections.

### PR DESCRIPTION
**Pull request by dcpleung**
If CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT is enabled,
the app shared memory partition may cause waste of memory
due to the need for padding.

For example, tests/subsys/jwt and board mps2_an385:

```
  z_test_mem_partition: addr 0x20000000, size 52
  z_libc_partition: addr 0x20000040, size 4
  k_mbedtls_partition : addr 0x20008000, size 32736

    ending at 0x2000ffff, taking up 65536 bytes
```

With power-of-two size and alignment requirement,
k_mbedtls_partition takes up 32KB memory and needs to be
aligned on 32KB boundary. If the above partitions are
ordered as shown, there needs to be a lot of padding
after z_libc_partition before k_mbedtls_partition can
start. In order to minimize padding, these partitions
need to be sort by size in descending order.

After the changes here, the partitions are:

```
  k_mbedtls_partition : addr 0x20000000, size 32736
  z_test_mem_partition: addr 0x20008000, size 52
  z_libc_partition: addr 0x20008040, size 4

    ending at 0x2000805f, taking up 32864 bytes
```

With the above example, sorting results in a saving
of 32672 bytes of saving.

Fixes #14121

### CODE REVIEW

**Comment by dcpleung** [comment]
recheck

**Comment by dcpleung** [comment]
I don't own the original `app_smem.ld`, so I am hesitant in adding copyright and license.

File: scripts/gen_app_partitions.py [hunk 1]
```diff
@@ -25,7 +32,7 @@
 """
 
 library_data_template = """
-		*{0}:*(.data .data.*)
+		*{0}:*(SORT_BY_ALIGNMENT(.data) SORT_BY_ALIGNMENT(.data.*))
```
Comment (andrewboie) [inline scripts/gen_app_partitions.py]: strictly speaking, if I understand correctly this change is orthogonal to this patch and is more related to #14183 can you make this change there instead?

File: scripts/gen_app_partitions.py [hunk 2]
```diff
@@ -34,7 +41,8 @@
 """
 
 library_bss_template = """
-		*{0}:*(.bss .bss.* COMMON COMMON.*)
+		*{0}:*(SORT_BY_ALIGNMENT(.bss) SORT_BY_ALIGNMENT(.bss.*)
+		       SORT_BY_ALIGNMENT(COMMON) SORT_BY_ALIGNMENT(COMMON.*))
```
Comment (andrewboie) [inline scripts/gen_app_partitions.py]: strictly speaking, if I understand correctly this change is orthogonal to this patch and is more related to #14183 can you make this change there instead?

**Review by ioannisg — COMMENTED** [review]
Looks very good. I am going to test it tomorrow.
One question; how is it done, now, with the subsections of app mem partitions (i.e. bss, data) - are they still adjacent sections, @andrewboie ?

**Comment by dcpleung** [comment]
Removed SORT_BY_ALIGNMENT(). Though I haven't add the copyright/license since the original copyright/license is not clear.

**Comment by andrewboie** [comment]
> One question; how is it done, now, with the subsections of app mem partitions (i.e. bss, data) - are they still adjacent sections, @andrewboie ?

Yes, should be. These are actually all in one output section for application shared memory. What the script now does is arrange the partitions in reverse order of alignment. But they still form blocks of data and then bss for each.

**Review by tyrantlizard77 — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Review by andyross — APPROVED** [review]
Idea is sound.  Python looks correct.  I have trouble reading cmake but I trust you.

**Comment by ruuddw** [comment]
Sorting the sections by alignement is a great improvement! Just one suggestion to consider:
 I understand it only makes a difference in case of power-of-two sized MPU regions, and hence a new option for the scripts and cmake is introduced to only do the alignement sorting for power-of-two MPUs. However, doing the alignment sorting wouldn't hurt in case of the arbitrary MPU size either, right? Simply always doing the alignment makes the cmake and scripting less complex. Each #ifdef and option we add increases maintenance/test/etc. costs...

**Comment by SebastianBoe** [comment]
@ruuddw : Correct. I assume that the option was added to allow users to change this in case any issues popped up. It would be best if we were able to drop the option.

**Comment by dcpleung** [comment]
From what I can tell from the map file, using SORT_BY_ALIGNMENT() or the linker command line option would not work in this case, since both options sort input section. For example,

```
datas () : ALIGN_WITH_INPUT
{
   *(.data)
   *(".data.*")
   *(".kernel.*")
} > RAM AT> ROM
```

becomes:

```
datas () : ALIGN_WITH_INPUT
{
   *(SORT_BY_ALIGNMENT(.data))
   *(SORT_BY_ALIGNMENT(".data.*"))
   *(SORT_BY_ALIGNMENT(".kernel.*"))
} > RAM AT> ROM
```

if you do it by hand or use the linker command line option. The sorting only applies to one input section. So in this case, just `(.data)` or `(.data.*)` but never across input sections. Also, the ordering is preserved, and that means `(.kernel.*)` would never go before `(.data)` or `(.data.*)`. This PR actually tries to re-order the section blocks (one block corresponding to one partition) by their size in descending order, where one block contains both bss and data for that partition.

**Review by nashif — APPROVED** [review]

**Comment by marc-hb** [comment]
PR #17032 broke this today, apologies. Fix submitted at PR #17132.
